### PR TITLE
Further parallelize index creation

### PR DESCRIPTION
**Pull request by taniabogatsch**
Before this PR, we sorted the index key columns and executed key column expressions in the `PhysicalCreateIndex` operator. However, this limited the sorting performance and decreased the index creation's performance in general.

This PR moves the sorting and the expression execution out of the `PhyscialCreateIndex` operator and adds them as extra operators to the CREATE INDEX pipeline. There are also some other minor code improvements.

### CODE REVIEW

File: src/execution/index/art/node256.cpp [hunk 1]
```diff
@@ -47,6 +47,16 @@ idx_t Node256::GetNextPos(idx_t pos) {
 	return Node::GetNextPos(pos);
 }
 
+idx_t Node256::GetNextPosAndByte(idx_t pos, uint8_t &byte) {
+	for (pos == DConstants::INVALID_INDEX ? pos = 0 : pos++; pos < 256; pos++) {
```
Comment (Tishj) [inline src/execution/index/art/node256.cpp:51]: Maybe a nitpick: can we use a `do while` here, or split this up somehow?
I think it's being too clever and it's a little hard to follow

File: src/execution/physical_plan/plan_create_index.cpp [hunk 2]
```diff
@@ -12,9 +14,27 @@ namespace duckdb {
...
 	D_ASSERT(op.children.empty());
 
+	// validate that all expressions contain valid scalar functions
+	// e.g. get_current_timestamp(), random(), and sequence values are not allowed as ART keys
+	// because they make deletions and lookups unfeasible
+	for (idx_t i = 0; i < op.unbound_expressions.size(); i++) {
+		auto &expr = op.unbound_expressions[i];
+		if (expr->expression_class == ExpressionClass::BOUND_FUNCTION) {
+			auto &func_expr = (BoundFunctionExpression &)*expr;
+			if (func_expr.function.name == "get_current_timestamp" || func_expr.function.name == "random" ||
```
Comment (Tishj) [inline src/execution/physical_plan/plan_create_index.cpp:29]: I wonder if this is too error-prone, since other extensions could also make functions, or supply a UDF
maybe BoundFunctionExpression needs a `IsDeterministic` flag of some similar?

Or maybe for now add a limit that the function has to come from the system catalog?

**Review by Tishj — COMMENTED** [review]
I made some nitpicky observations, but it looks good :)

One more thought I had, but I'm not sure of the added benefit over the existing solution:
Maybe we want to create a NodeChildIterator instead of the NextPosAndByte functions?
That would encapsulate the position, so you don't have to keep track of this position yourself and keep giving it back to the function.

File: src/execution/index/art/node256.cpp (hunk 1 above)
Comment (taniabogatsch) [reply src/execution/index/art/node256.cpp:51]: I have a separate PR after this one to replace all the magic numbers (this is not the only time in the ART code haha), it is part of the track-index-size refactor.

Comment (taniabogatsch) [reply src/execution/index/art/node256.cpp:51]: If that works for you, I'd like to leave it like this for now, and I'll add these changes to my other branch avoiding merge conflicts for all the `nodeXX.cpp` files?

File: src/execution/physical_plan/plan_create_index.cpp (hunk 2 above)
Comment (taniabogatsch) [reply src/execution/physical_plan/plan_create_index.cpp:29]: Turns out that someone already thought of this! :) 
```c++
enum class FunctionSideEffects : uint8_t { NO_SIDE_EFFECTS = 0, HAS_SIDE_EFFECTS = 1 };
...
//! Whether or not the function has side effects (e.g. sequence increments, random() functions, NOW()). Functions
//! with side-effects cannot be constant-folded.
FunctionSideEffects side_effects;
```

**Review by Tishj — APPROVED** [review]
Looks good to me :)

File: test/fuzzer/pedro/index_current_timestamp.test [hunk 3]
```diff
@@ -13,38 +13,31 @@ INSERT INTO t1(c1) VALUES(1);
...
-statement error
+statement ok
 INSERT INTO t1 VALUES (42, 'hello')
-----
-Cannot use get_current_timestamp
 
 restart
 
 statement error
+CREATE INDEX i1 ON t1((get_current_timestamp()), c1);
```
Comment (Tmonster) [inline test/fuzzer/pedro/index_current_timestamp.test:40]: nit: I believe there is a way now to check the error returned. Maybe you have this in another PR though?

Comment (taniabogatsch) [reply test/fuzzer/pedro/index_current_timestamp.test:40]: Oh that is very neat to know, thanks!

File: src/execution/physical_plan/plan_create_index.cpp [hunk 4]
```diff
@@ -12,9 +14,26 @@ namespace duckdb {
...
+	// table scan - projection (for expression execution) - filter (NOT NULL) - order - create index
+
 	D_ASSERT(op.children.empty());
 
+	// validate that all expressions contain valid scalar functions
+	// e.g. get_current_timestamp(), random(), and sequence values are not allowed as ART keys
+	// because they make deletions and lookups unfeasible
+	for (idx_t i = 0; i < op.unbound_expressions.size(); i++) {
+		auto &expr = op.unbound_expressions[i];
+		if (expr->expression_class == ExpressionClass::BOUND_FUNCTION) {
```
Comment (Mytherin) [inline src/execution/physical_plan/plan_create_index.cpp:27]: Can we use `expr->HasSideEffects()` here? That also checks this property recursively in the children of the expression.

**Review by Mytherin — COMMENTED** [review]
Thanks! Just one minor comment - perhaps you can fix it in a future PR:

File: src/execution/physical_plan/plan_create_index.cpp (hunk 4 above)
Comment (taniabogatsch) [reply src/execution/physical_plan/plan_create_index.cpp:27]: Sure, I added a reminder to #5865.
